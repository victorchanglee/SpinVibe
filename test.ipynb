{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Hs1' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 13\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[38;5;66;03m# Displacements (x, y) and corresponding H_s values\u001b[39;00m\n\u001b[1;32m     10\u001b[0m displacements \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.1\u001b[39m, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.1\u001b[39m], [\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.1\u001b[39m, \u001b[38;5;241m0.0\u001b[39m], [\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.1\u001b[39m, \u001b[38;5;241m0.1\u001b[39m],\n\u001b[1;32m     11\u001b[0m                           [\u001b[38;5;241m0.0\u001b[39m, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.1\u001b[39m], [\u001b[38;5;241m0.0\u001b[39m, \u001b[38;5;241m0.0\u001b[39m], [\u001b[38;5;241m0.0\u001b[39m, \u001b[38;5;241m0.1\u001b[39m],\n\u001b[1;32m     12\u001b[0m                           [\u001b[38;5;241m0.1\u001b[39m, \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m0.1\u001b[39m], [\u001b[38;5;241m0.1\u001b[39m, \u001b[38;5;241m0.0\u001b[39m], [\u001b[38;5;241m0.1\u001b[39m, \u001b[38;5;241m0.1\u001b[39m]])\n\u001b[0;32m---> 13\u001b[0m Hs_values \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray([\u001b[43mHs1\u001b[49m, Hs2, Hs3, Hs4, Hs5, Hs6, Hs7, Hs8, Hs9])  \u001b[38;5;66;03m# Replace with actual values\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# Fit the polynomial to the data\u001b[39;00m\n\u001b[1;32m     16\u001b[0m coords \u001b[38;5;241m=\u001b[39m (displacements[:, \u001b[38;5;241m0\u001b[39m], displacements[:, \u001b[38;5;241m1\u001b[39m])  \u001b[38;5;66;03m# x and y displacements\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Hs1' is not defined"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from scipy.optimize import curve_fit\n",
    "\n",
    "# Define the polynomial function\n",
    "def poly_func(coords, a0, a1, a2, a3, a4, a5):\n",
    "    x, y = coords\n",
    "    return a0 + a1*x + a2*y + a3*x**2 + a4*y**2 + a5*x*y\n",
    "\n",
    "# Displacements (x, y) and corresponding H_s values\n",
    "displacements = np.array([[-0.1, -0.1], [-0.1, 0.0], [-0.1, 0.1],\n",
    "                          [0.0, -0.1], [0.0, 0.0], [0.0, 0.1],\n",
    "                          [0.1, -0.1], [0.1, 0.0], [0.1, 0.1]])\n",
    "Hs_values = np.array([Hs1, Hs2, Hs3, Hs4, Hs5, Hs6, Hs7, Hs8, Hs9])  # Replace with actual values\n",
    "\n",
    "# Fit the polynomial to the data\n",
    "coords = (displacements[:, 0], displacements[:, 1])  # x and y displacements\n",
    "popt, _ = curve_fit(poly_func, coords, Hs_values)\n",
    "\n",
    "# Extract coefficients\n",
    "a0, a1, a2, a3, a4, a5 = popt\n",
    "\n",
    "# Compute second derivatives\n",
    "d2Hs_dx2 = 2 * a3\n",
    "d2Hs_dy2 = 2 * a4\n",
    "d2Hs_dxdy = a5\n",
    "\n",
    "print(f\"Second derivative with respect to x^2: {d2Hs_dx2}\")\n",
    "print(f\"Second derivative with respect to y^2: {d2Hs_dy2}\")\n",
    "print(f\"Mixed second derivative (dxdy): {d2Hs_dxdy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1D Array 1:\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "1D Array 2:\n",
      "[10 11 12 13 14 15 16 17 18 19]\n",
      "\n",
      "2D Array (10x10):\n",
      "[[10 11 12 13 14 15 16 17 18 19]\n",
      " [11 12 13 14 15 16 17 18 19 20]\n",
      " [12 13 14 15 16 17 18 19 20 21]\n",
      " [13 14 15 16 17 18 19 20 21 22]\n",
      " [14 15 16 17 18 19 20 21 22 23]\n",
      " [15 16 17 18 19 20 21 22 23 24]\n",
      " [16 17 18 19 20 21 22 23 24 25]\n",
      " [17 18 19 20 21 22 23 24 25 26]\n",
      " [18 19 20 21 22 23 24 25 26 27]\n",
      " [19 20 21 22 23 24 25 26 27 28]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Define two 1D arrays of length 10\n",
    "array1 = np.arange(10)\n",
    "array2 = np.arange(10, 20)\n",
    "\n",
    "# Create a 2D array of size 10x10\n",
    "# Using broadcasting to combine the two arrays into a 2D array\n",
    "array_2d = np.array([array1 + i for i in array2])\n",
    "\n",
    "# Print the 2D array\n",
    "print(\"1D Array 1:\")\n",
    "print(array1)\n",
    "print(\"\\n1D Array 2:\")\n",
    "print(array2)\n",
    "print(\"\\n2D Array (10x10):\")\n",
    "print(array_2d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1D Array of 2D Vectors:\n",
      "[[ 0 10]\n",
      " [ 1 11]\n",
      " [ 2 12]\n",
      " [ 3 13]\n",
      " [ 4 14]\n",
      " [ 5 15]\n",
      " [ 6 16]\n",
      " [ 7 17]\n",
      " [ 8 18]\n",
      " [ 9 19]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Define two 1D arrays of length 10\n",
    "array1 = np.arange(10)\n",
    "array2 = np.arange(10, 20)\n",
    "\n",
    "# Create a 1D array of 2D vectors (each 2D vector is a NumPy array with 2 elements)\n",
    "array_1d_of_2d_vectors = np.array([[array1[i], array2[i]] for i in range(len(array1))])\n",
    "\n",
    "# Print the 1D array of 2D vectors\n",
    "print(\"1D Array of 2D Vectors:\")\n",
    "print(array_1d_of_2d_vectors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1D Array 1:\n",
      "[0 1 2 3 4 5 6 7 8 9]\n",
      "\n",
      "1D Array 2:\n",
      "[10 11 12 13 14 15 16 17 18 19]\n",
      "\n",
      "Cartesian Product (100 pairs):\n",
      "[[ 0 10]\n",
      " [ 0 11]\n",
      " [ 0 12]\n",
      " [ 0 13]\n",
      " [ 0 14]\n",
      " [ 0 15]\n",
      " [ 0 16]\n",
      " [ 0 17]\n",
      " [ 0 18]\n",
      " [ 0 19]\n",
      " [ 1 10]\n",
      " [ 1 11]\n",
      " [ 1 12]\n",
      " [ 1 13]\n",
      " [ 1 14]\n",
      " [ 1 15]\n",
      " [ 1 16]\n",
      " [ 1 17]\n",
      " [ 1 18]\n",
      " [ 1 19]\n",
      " [ 2 10]\n",
      " [ 2 11]\n",
      " [ 2 12]\n",
      " [ 2 13]\n",
      " [ 2 14]\n",
      " [ 2 15]\n",
      " [ 2 16]\n",
      " [ 2 17]\n",
      " [ 2 18]\n",
      " [ 2 19]\n",
      " [ 3 10]\n",
      " [ 3 11]\n",
      " [ 3 12]\n",
      " [ 3 13]\n",
      " [ 3 14]\n",
      " [ 3 15]\n",
      " [ 3 16]\n",
      " [ 3 17]\n",
      " [ 3 18]\n",
      " [ 3 19]\n",
      " [ 4 10]\n",
      " [ 4 11]\n",
      " [ 4 12]\n",
      " [ 4 13]\n",
      " [ 4 14]\n",
      " [ 4 15]\n",
      " [ 4 16]\n",
      " [ 4 17]\n",
      " [ 4 18]\n",
      " [ 4 19]\n",
      " [ 5 10]\n",
      " [ 5 11]\n",
      " [ 5 12]\n",
      " [ 5 13]\n",
      " [ 5 14]\n",
      " [ 5 15]\n",
      " [ 5 16]\n",
      " [ 5 17]\n",
      " [ 5 18]\n",
      " [ 5 19]\n",
      " [ 6 10]\n",
      " [ 6 11]\n",
      " [ 6 12]\n",
      " [ 6 13]\n",
      " [ 6 14]\n",
      " [ 6 15]\n",
      " [ 6 16]\n",
      " [ 6 17]\n",
      " [ 6 18]\n",
      " [ 6 19]\n",
      " [ 7 10]\n",
      " [ 7 11]\n",
      " [ 7 12]\n",
      " [ 7 13]\n",
      " [ 7 14]\n",
      " [ 7 15]\n",
      " [ 7 16]\n",
      " [ 7 17]\n",
      " [ 7 18]\n",
      " [ 7 19]\n",
      " [ 8 10]\n",
      " [ 8 11]\n",
      " [ 8 12]\n",
      " [ 8 13]\n",
      " [ 8 14]\n",
      " [ 8 15]\n",
      " [ 8 16]\n",
      " [ 8 17]\n",
      " [ 8 18]\n",
      " [ 8 19]\n",
      " [ 9 10]\n",
      " [ 9 11]\n",
      " [ 9 12]\n",
      " [ 9 13]\n",
      " [ 9 14]\n",
      " [ 9 15]\n",
      " [ 9 16]\n",
      " [ 9 17]\n",
      " [ 9 18]\n",
      " [ 9 19]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "# Define two 1D arrays of length 10\n",
    "array1 = np.arange(10)\n",
    "array2 = np.arange(10, 20)\n",
    "\n",
    "# Generate the Cartesian product of array1 and array2\n",
    "cartesian_product = list(itertools.product(array1, array2))\n",
    "\n",
    "# Convert the list of tuples to a NumPy array\n",
    "cartesian_product_array = np.array(cartesian_product)\n",
    "\n",
    "# Print the result\n",
    "print(\"1D Array 1:\")\n",
    "print(array1)\n",
    "print(\"\\n1D Array 2:\")\n",
    "print(array2)\n",
    "print(\"\\nCartesian Product (100 pairs):\")\n",
    "print(cartesian_product_array)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "work",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
