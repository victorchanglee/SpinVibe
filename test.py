import numpy as np

def zeeman_term_per_spin(B, g_tensors, spins, beta):
    """
    Compute the Zeeman interaction energy per spin.
    """
    zeeman_energies = []
    for i, spin in enumerate(spins):
        energy = beta[i] * np.dot(B, np.dot(g_tensors[i], spin))
        zeeman_energies.append(energy)
    return np.array(zeeman_energies)

def crystal_field_term_per_spin(B_lm, spin_operators, spins):
    """
    Compute the crystal field interaction energy per spin.
    """
    crystal_field_energies = []
    for i, spin in enumerate(spins):
        energy = 0.0
        S_magnitude = np.linalg.norm(spin)
        S_operators = spin_operators[i]
        for l in range(2, int(2 * S_magnitude) + 1):
            for m in range(-l, l + 1):
                energy += B_lm[i][l][m] * S_operators[(l, m)]
        crystal_field_energies.append(energy)
    return np.array(crystal_field_energies)

def exchange_term_per_spin(spins, J_tensors):
    """
    Compute the exchange interaction energy per spin.
    """
    N = len(spins)
    exchange_energies = np.zeros(N)
    for i in range(N):
        for j in range(N):
            if i != j:
                exchange_energies[i] += np.dot(spins[i], np.dot(J_tensors[i, j], spins[j]))
    return exchange_energies

def compute_hamiltonian_per_spin(B, g_tensors, spins, beta, B_lm, spin_operators, J_tensors):
    """
    Compute the Hamiltonian contributions per spin.
    """
    # Zeeman interaction term per spin
    zeeman_energies = zeeman_term_per_spin(B, g_tensors, spins, beta)

    # Crystal field term per spin
    crystal_field_energies = crystal_field_term_per_spin(B_lm, spin_operators, spins)

    # Exchange interaction term per spin
    exchange_energies = exchange_term_per_spin(spins, J_tensors)

    # Total energy per spin
    total_energies = zeeman_energies + crystal_field_energies + 0.5 * exchange_energies
    return total_energies

# Example Inputs
B = np.array([0.1, 0.2, 0.3])  # Magnetic field vector
g_tensors = [np.eye(3) for _ in range(3)]  # Identity g-tensors for 3 spins
spins = [np.array([1.0, 0.0, 0.0]), np.array([0.0, 1.0, 0.0]), np.array([0.0, 0.0, 1.0])]  # Spin vectors
beta = [1.0, 1.5, 2.0]  # Coupling constants

# Crystal field coefficients and operators
B_lm = [{l: {m: 0.1 for m in range(-l, l + 1)} for l in range(2, 4)} for _ in range(3)]
spin_operators = [{(l, m): 0.1 for l in range(2, 4) for m in range(-l, l + 1)} for _ in range(3)]

# Exchange coupling tensors
J_tensors = np.zeros((3, 3, 3, 3))  # Example 3x3 spin system
J_tensors[0, 1] = np.eye(3)  # Interaction between spin 0 and 1
J_tensors[1, 2] = np.eye(3)  # Interaction between spin 1 and 2

# Compute the Hamiltonian energy per spin
H_s_per_spin = compute_hamiltonian_per_spin(B, g_tensors, spins, beta, B_lm, spin_operators, J_tensors)

print(f"Hamiltonian energy per spin: {H_s_per_spin}")