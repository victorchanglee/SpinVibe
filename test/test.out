Reading input parameters from: input.json
Initializing input files...
  spin_file: Cr_otolyl_4.h5
  phonon_file: phonon.h5
  d1_file: Cr_otolyl_4_d1.h5
  d2_file: d_16.h5
  g2_file: g_16.h5
  atoms_file: Sn_otolyl_4_atoms.h5
  indices_file: molecule_indices.h5
  mol_mass: mol_mass.h5
2025-10-10 13:39:19
---------------------------------------
|Start Spin-phonon coupling simulation|
---------------------------------------


Input Parameters
================
Magnetic field: [0.   0.   0.41]
S: 1
T: 5.0
Number of unit cells: 50
Broadening: 1.0
Population type: pure
Polarization: [0 0 1]
Rotational matrix:
[[0 0 1]
 [1 0 0]
 [0 1 0]]


Spin Hamiltonian
[[-3.61376628e-01+0.00000000e+00j -2.91812008e-06+1.06333248e-05j
   5.11553213e-02-7.05205722e-06j]
 [-2.91812008e-06-1.06333248e-05j -3.41022461e-02+0.00000000e+00j
   2.91812008e-06-1.06333248e-05j]
 [ 5.11553213e-02+7.05205722e-06j  2.91812008e-06+1.06333248e-05j
   3.95478874e-01+0.00000000e+00j]]


Eigenvalues of the spin Hamiltonian
[-0.36481853 -0.03410225  0.39892077]


Initialize simulation
=====================
Computing linear and quadratic coupling
Running in serial mode.


Initializing R1 tensor (Linear coupling)
Eigenvalues of the R1 matrix
[-7.58667475e+03+2.29084924e-13j -6.38860182e+03+2.04917103e-13j
 -4.43206606e+03-6.17862218e-13j -3.59220371e+03+3.20963000e-13j
  5.08795295e-12+6.61490434e-14j -6.82711695e+02-1.48816311e-13j
 -1.01603012e+03+1.59702903e-13j -1.55496452e+03+2.75645369e-13j
 -1.89462756e+03+2.49180632e-13j]


Initializing R2 tensor (Quadratic coupling)
Number of tasks per process: 58653
Eigenvalues of the R2 matrix
[-7.64665795e+02-3.41017720e-14j -6.28162322e+02+2.53751893e-15j
 -5.88361088e+02+2.14493946e-14j -5.40094994e+02-4.00574650e-14j
  3.32044503e-12+3.17624404e-15j -2.26146073e+02-4.74845855e-15j
 -1.34649707e+02+7.87856726e-14j -1.05801324e+02+1.01395004e-14j
 -1.70499037e+02+3.03209249e-14j]


Eigenvalues of the Redfield matrix
[-8.32645270e+03+4.79372864e-14j -6.96442342e+03-2.02997631e-13j
 -5.00639954e+03+1.88747491e-13j -4.06763728e+03+2.49690690e-13j
  8.22548089e-12+2.45447719e-13j -8.03140622e+02-9.38716507e-14j
 -1.20061821e+03-2.36810370e-14j -1.82002065e+03+2.19803262e-13j
 -2.11756815e+03-5.74232710e-13j]


Initial spin population:
[[0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 0.+0.j]
 [0.+0.j 0.+0.j 1.+0.j]]


Start time evolution


Equilibrium M_parallel: 0.07609010766145541
Initial Magnetization: 1.0000
Fitted Equilibrium Magnetization: 0.0761
RÂ² = 0.9983
T1 from magnetization decay
T1 =  0.00024210452040129466 s
T1_err =  8.287624090381662e-07 s


Saving data
Data has been saved to Spin_phonon_5K_1.0_50_0.0.h5


Initiate simulation: 0h 0m 42.77s
Build R1: 0h 0m 0.57s
Build R2: 0h 1m 54.52s
Time evolution: 0h 0m 0.01s
Measuring Time: 0h 0m 0.15s
Total Run Time: 0h 2m 38.04s
Job completed. Results saved to Spin_phonon_5K_1.0_50_0.0.h5
